### ISSUE TITLE
Sync metadata has n+1 on some drivers

### ISSUE THREAD

**Issue by qnkhuat** [metabase/metabase#37742]
During the sync process, we have the method `driver/describe-database` to find all the tables within a database. For some drivers we implemented so that we only sync tables that the current connection has select privilege against.

The current checking method is executing a query `select true from table limit 1` see `(metabase.driver.sql-jdbc.sync.interface/have-select-privilege?)`, this is inefficient and is a basic N+1 problem.

We have fixed this for Postgres, redshift, and MySQL in #37439 by using the output of `driver/current-user-table-privileges` to filter out tables where we don't have select privileges. 

This is a much better approach, so we should extend it to supports: sparksql, oracle, snowflake, athena, sqlite, maraidb, sqlserver.
For these we can implement `driver/current-user-table-privileges` and they will automatically be fast.

Additionally, we have some drivers that do not check select privileges during sync, like bigquery, vertica, or presto; we should update so that those drivers doing this check, too.

**Comment by paoliniluis** [comment]
@qnkhuat can we implement this for snowflake as well?

**Comment by qnkhuat** [comment]
yes we can, we just need to figure out how to to properly getting snowflake privileges info

**Comment by petrenkorf** [comment]
@qnkhuat May I take this task?

**Comment by qnkhuat** [comment]
yes, go for it!

**Comment by dpsutton** [comment]
Examples of implementations:
- postgres: https://github.com/metabase/metabase/blob/master/src/metabase/driver/postgres.clj#L1229-L1259
- redshift (disabled): https://github.com/metabase/metabase/blob/master/modules/drivers/redshift/src/metabase/driver/redshift.clj#L624
- mysql: https://github.com/metabase/metabase/blob/master/src/metabase/driver/mysql.clj#L224-L241

**Comment by petrenkorf** [comment]
@qnkhuat @dpsutton What drivers are pending the implementation? I am planning to take a look at this soon.
